plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,316 @@
+      description = "Grammar packages to install";
+    };
+
+    languageRegister = mkOption {
+      type = with types; attrsOf (coercedTo str toList (listOf str));
+      default = { };
+      example = {
+        cpp = "onelab";
+        python = [
+          "foo"
+          "bar"
+        ];
       };
+      description = ''
+        This is a wrapping of the `vim.treesitter.language.register` function.
+        Register specific parsers to one or several filetypes.
+        The keys are the parser names and the values are either one or several filetypes.
+      '';
+    };
 
-      nixvimInjections = mkEnableOption "nixvim specific injections, like lua highlighting in extraConfigLua";
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = "Whether to automatically install grammar packages.";
+    };
+
+    nixvimInjections = mkOption {
+      type = types.bool;
+      default = true;
+      example = false;
+      description = ''Whether to enable Nixvim injections, such as recognising extraConfigLua as lua.'';
+    };
+
+    nodejsPackage = helpers.mkPackageOption {
+      name = "nodejs";
+      default = pkgs.nodejs;
+      defaultText = literalExpression "pkgs.nodejs";
+      example = literalExpression "pkgs.nodejs_22";
+      description = ''
+        Which package (if any) to be added as the nodejs package.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of NodeJS, set this option to `null`.
+      '';
+    };
+
+    treesitterPackage = helpers.mkPackageOption {
+      name = "treesitter";
+      default = pkgs.tree-sitter;
+      defaultText = literalExpression "pkgs.treesitter";
+      description = ''
+        Which package (if any) to be added as the treesitter binary.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of treesitter, set this option to `null`.
+      '';
     };
   };
 
-  config =
-    let
-      tsOptions = {
-        highlight = {
-          inherit (cfg) enable;
-          disable = if (cfg.disabledLanguages != [ ]) then cfg.disabledLanguages else null;
+  # NOTE: We call require('nvim-treesitter.configs').setup in extraConfigLua.
+  callSetup = false;
+  # NOTE: We dont want to try installing the provided package because we install it in extraPlugins
+  installPackage = false;
 
-          custom_captures = if (cfg.customCaptures != { }) then cfg.customCaptures else null;
-        };
+  extraConfig = cfg: {
+    assertions = [
+      {
+        assertion = ((cfg.grammarPackages != [ ]) || cfg.grammarPackages == "all") -> cfg.package != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have enabled `grammarPackages`, but `plugins.treesitter.package` is `null`.
+          This package is required to choose which grammars to install.
+        '';
+      }
```

Comment:
Turns out cfg.package isn't nullable atm, so we can drop this assertion.

@GaetanLepage unless you think we should keep it in case we change `package`'s type to be nullable?